### PR TITLE
Split Include into Include and RawInclude

### PR DESCRIPTION
**Pull request by TimothyGu**

### CODE REVIEW

**Comment by ForbesLindesay** [comment]
We should be consistent between handling of single vs. multi level filtered include.  Either:
1. Filtered include has an array of filter name + attributes
2. Filter node can contain either plain text or an include node, there is no such thing as a "filtered include"

My proposal would be to have a `RawInclude` node which would be for including plain text, and an `Include` node that is for including jade.  Then as per option 2, a filter node can contain a RawInclude or another Filter or plain text

**Comment by TimothyGu** [comment]
> Filter node can contain either plain text or an include node, there is no such thing as a "filtered include"

That was my initial plan, but I wanted to be able to use JSTransformer's renderFile, and a separate type for filtered include seemed to be a natural solution.

**Comment by ForbesLindesay** [comment]
That's a good point, I suppose our handling of the first filter will be special in that respect, although it seems like an implementation detail of the `filter` processing step (i.e. it could look at the node, see if it's a filter node, then see if it's child is an include node, then behave appropriately).

**Comment by ForbesLindesay** [comment]
Maybe go with option 1 then?  It will make it easier to implement efficiently without generating and then discarding a whole series of plain text nodes.

**Comment by TimothyGu** [comment]
How about this:
- `RawInclude` for raw inclusion; `.filters` can be `[ IncludeFilter ] | null`; `IncludeFilter` is just like a regular `Filter` but without `block`; filters are applied in ascending order from `.filters[0]`.
- `Include` for Jade inclusion.

**Comment by TimothyGu** [comment]
@ForbesLindesay ping.

**Comment by ForbesLindesay** [comment]
Lets have `.filters` as the empty array instead of `null` when there are no filters.  Other than that I'm completely happy with this.

**Comment by TimothyGu** [comment]
Updated, please review.

**Comment by TimothyGu** [comment]
One more problem with this approach: the exact path and therefore type of the inclusion is determined in jade-load's overridable resolve function, which means it is not possible to tell if the inclusion is RawInclude or Include at the parsing stage, unless other factors like filters are involved.

So this leaves us with two choices:
- Add `.jade` suffix at the parsing stage and disallow overriding `resolve` in jade-load. In other words,  Include is used for all extensionless includes (again, unless there are filters). Changing the parser is of course always possible, but not as convenient.
- Use filters as the sole determinant of the type of the inclusion, i.e. we are back to the original FilteredInclude type.

**Comment by ForbesLindesay** [comment]
I think we should disallow changing the type at the `jade-load` stage.  i.e. if the node is `RawInclude` then we include it as plain text even if the file extension is `.jade`.  It's then up to the parser to decide whether something is Include or RawInclude.

In short, lets deprecate extension-less include (and extend)

File: SPEC.md [hunk 1]
```diff
@@ -266,14 +266,36 @@ A node with file reference information.
 ### Include
 
 ```js
-interface Include <: BlockNode, FileNode {
+interface IncludeNode <: BlockNode, FileNode {}
```
Comment (ForbesLindesay) [inline SPEC.md:269]: I don't think `RawInclude` should be a `BlockNode`?  Perhaps we should remove `IncludeNode` and just have `Include` and `RawInclude` directly extend `FileNode`.

Comment (TimothyGu) [inline SPEC.md:269]: Oops…

**Pull request merged**